text/0000-dual-light-shadow.md

```diff
@@ -0,0 +1,297 @@
+
+Also, the design of LWC lends itself to largely shadow-DOM-neutral component authoring. For instance, LWC supports scoped styles, light DOM `<slot>`s, and `lwc:ref` – all of which work roughly the same between shadow components and light components. So there is not a strong reason for a component to declare upfront whether it is light or shadow.
+
+So naturally, a request we've heard from component authors is that they would like to be able to author a single component and have it render as either light or shadow DOM.
+
+To be fair, there is already a workaround today: composition. I.e., a component could be authored in light DOM, and then a wrapper component could be authored in shadow DOM.
+
+However, this workaround has several downsides:
+
+1. All properties have to be reflected between the shadow parent and light child.
+2. All slots have to be propagated from the parent to the child.
+3. An extra wrapper component is required, which may be undesired for common base components (e.g. buttons, icons).
+4. The two components have different external tag names (e.g. `my-button` and `my-button-shadow`).
+
+This RFC has none of these downsides.
+
+## Detailed design
+
+### Declaring a dual-mode component
+
+Currently the `lwc:render-mode` directive / `static renderMode` property only allows two values: `"light"` and `"shadow"`.
+
+This proposal adds a third value: `"dual"`.
+
+```html
+<!-- component.html -->
+<template lwc:render-mode="dual">
+    <h1>Hello world</h1>
+</template>
+```
+
+```js
+// component.js
+export default class extends LightningElement {
+  static renderMode = 'dual';
+}
+```
+
+### Switching between light and shadow mode
+
+Currently, `lwc:render-mode` can only be applied to the top-level `<template>`. In this proposal,
+it can also be applied to LWC components referenced inside a template:
+
+```html
+<template>
+    <x-component lwc:render-mode="light"></x-component>
+    <x-component lwc:render-mode="shadow"></x-component>
+</template>
+```
+
+When used in this way, the value of `lwc:render-mode` can only be a static string, and it can only be `"light"` or `"shadow"`.
+
+If `"dual"` is used in this context, a compile-time error is thrown:
+
+```html
+<template>
+    <!-- Invalid -->
+    <x-component lwc:render-mode="dual"></x-component>
+</template>
+```
+
+If a template is declared to be `lwc:render-mode="dual"`, then the corresponding component must also have
+`static renderMode = 'dual'`, and vice versa.
+
+If not, an error will be thrown at runtime. (This is the same as what currently happens if `"shadow"` and `"light"`
+are mixed between the two.)
+
+### Non-LWC components
+
+`lwc:render-mode` can only be applied to LWC components. If it's applied to any other element, including an 
+`lwc:external` component, then a compile-time error is thrown:
+
+```html
+<template>
+    <!-- Invalid -->
+    <third-party lwc:render-mode="light"></third-party>

```

Comment:
How do you know at built-time that this is a non-LWC? You don't. Is this example missing the `lwc:external` directive?

Reply:
Sorry, yes. I missed the `lwc:external` here.